applications/StructuralMechanicsApplication/custom_response_functions/response_utilities/adjoint_nonlinear_strain_energy_response_function.cpp

```diff
@@ -0,0 +1,385 @@
+    {
+        KRATOS_TRY;
+
+        ProcessInfo process_info = rProcessInfo;
+
+        const SizeType number_of_nodes = rAdjointCondition.GetGeometry().size();
+        const SizeType dimension = rAdjointCondition.GetGeometry().WorkingSpaceDimension();
+        const SizeType vec_size = number_of_nodes * dimension;
+
+        Vector displacement = ZeroVector(vec_size);
+        Vector displacement_previous_step = ZeroVector(vec_size);
+        double delta = rAdjointCondition.GetValue(PERTURBATION_SIZE);
+        Vector RHS;
+
+        // getting the displacements for the current and the previous time steps
+        int i = 0;
+        for (auto& node_i : rAdjointCondition.GetGeometry())
+        {
+            project(displacement, range(i, dimension)) = rAdjointCondition.GetGeometry()[node_i].FastGetSolutionStepValue(DISPLACEMENT);
+            project(displacement_previous_step , range(i , dimension))= rAdjointCondition.GetGeometry()[node_i].FastGetSolutionStepValue(DISPLACEMENT,1);
+            i += 3;
+        }
+
+        mConditions[rAdjointCondition.Id()]->CalculateRightHandSide(RHS, process_info);
+
+        //calculation of partial f_{ext}_i}{\partial u}
+        Matrix partial_derivative_matrix = ZeroMatrix(vec_size, vec_size);
+        int i_2 = 0;
+        for (auto& node_i : rAdjointCondition.GetGeometry())
+        {
+            Vector perturbed_RHS = Vector(0);
+
+            // Pertubation, gradient analysis and recovery of x
+            node_i.X() += delta;
+            node_i.FastGetSolutionStepValue(DISPLACEMENT_X) += delta;
+            mConditions[rAdjointCondition.Id()]->CalculateRightHandSide(perturbed_RHS, process_info);
+            row(partial_derivative_matrix, i_2) = (perturbed_RHS - RHS) / delta;
+            node_i.X() -= delta;
+            node_i.FastGetSolutionStepValue(DISPLACEMENT_X) -= delta;
+
+            // Reset the pertubed vector
+            perturbed_RHS = Vector(0);
+
+            // Pertubation, gradient analysis and recovery of y
+            node_i.Y() += delta;
+            node_i.FastGetSolutionStepValue(DISPLACEMENT_Y) += delta;
+            mConditions[rAdjointCondition.Id()]->CalculateRightHandSide(perturbed_RHS, process_info);
+            row(partial_derivative_matrix, i_2 + 1) = (perturbed_RHS - RHS) / delta;
+            node_i.Y() -= delta;
+            node_i.FastGetSolutionStepValue(DISPLACEMENT_Y) -= delta;
+
+            // Reset the pertubed vector
+            perturbed_RHS = Vector(0);
+
+            // Pertubation, gradient analysis and recovery of z
+            node_i.Z() += delta;
+            node_i.FastGetSolutionStepValue(DISPLACEMENT_Z) += delta;
+            mConditions[rAdjointCondition.Id()]->CalculateRightHandSide(perturbed_RHS, process_info);
+            row(partial_derivative_matrix, i_2 + 2) = (perturbed_RHS - RHS) / delta;
+            node_i.Z() -= delta;
+            node_i.FastGetSolutionStepValue(DISPLACEMENT_Z) -= delta;
```

Comment:
In the future there could be also corresponding ROTATION dofs. @mahmoudsesa and @armingeiser we should discuss in person if we should consider this already at this stage.

Reply:
I think rotation is not needed because the external loads are not rotation dependent.

Reply:
In general there can be corresponding rotation dofs. Think e.g. on the element load vector of a line load on a beam.

Reply:
I will discuss that with you in person